river/sketch/set.py

```diff
@@ -0,0 +1,259 @@
+import math
+import random
+import typing
+
+from river import base
+
+
+class Set(base.Base):
+    """Approximate tracking of observed items using Bloom filters.
+
+    Bloom filters enable using a limited amount of memory to check whether a given item was already
+    observed in a stream. They can be used similarly to Python's built-in sets with the difference
+    that items are not explicitly stored. For that reason, element removal and set difference are not
+    currently supported.
+
+    Bloom filters store a bit array and map incoming items to `k` index positions in the such array.
+    The selected positions are set to `True`. Therefore, a binary code representation is created for each item.
+    Membership works by projecting the query item and checking if every position of its binary code is
+    `True`. If that is not the case, the item was not observed yet. A nice property of Bloom filters is
+    that they do not yield false negatives: unobserved items might be signalized as observed, but observed
+    items are never signalized as unobserved.
+
+    If more than one item has the same binary code, i.e., hash collisions happen, the accuracy of the Bloom
+    filter decreases, and false positives are produced. For instance, a previously unobserved item is signalized
+    as observed. Increasing the size of the binary array and the value of `k` increase the filter's accuracy as
+    hash collisions are avoided. Nonetheless, even using an increased number of hash functions, hash collisions
+    will frequently happen if the array capacity is too small. The length of the bit array and the number of
+    hash functions are inferred automatically from the supplied `capacity` and `fp_rate`.
+
+
+    Parameters
+    ----------
+    capacity
+        The maximum capacity of the Bloom filter, i.e., the maximum number of distinct items to store given
+        the selected `fp_rate`.
+    fp_rate
+        The allowed rate of false positives. The probability of obtaining a true positive is `1 - fp_rate`.
+    seed
+        Random seed for reproducibility.
+
+    Examples
+    --------
+
+    >>> import random
+    >>> from river import sketch
+
+    >>> rng = random.Random(42)
+    >>> s_set = sketch.Set(capacity=100, seed=0)
+
+
+    We can retrieve the number of selected hash functions:
+    >>> s_set.n_hash
+    7
+
+    And the size of the binary array used by the Bloom filter:
```

Comment:
I would add carriage return here too. I didn't comment at every sentence, but you get the idea :). Basically one carriage return between each sentence and each code block

Reply:
I see! Should I add one carriage return just after a code block or before too?